src/__tests__/waitForElement.js

```diff
@@ -0,0 +1,278 @@
+  expect(testEl.parentNode).toBe(container)
+
+  await promise
+})
+
+test('it waits for the next DOM mutation with default callback', async () => {
+  const successHandler = jest.fn().mockName('successHandler')
+  const errorHandler = jest.fn().mockName('errorHandler')
+
+  const promise = waitForElement().then(successHandler, errorHandler)
+
+  // Promise callbacks are always asynchronous.
+  expect(successHandler).toHaveBeenCalledTimes(0)
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+
+  await skipSomeTimeForMutationObserver()
+
+  // No more expected calls without DOM mutations.
+  expect(successHandler).toHaveBeenCalledTimes(0)
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+
+  document.body.appendChild(document.createElement('div'))
+  expect(document.body).toMatchSnapshot()
+
+  await skipSomeTimeForMutationObserver()
+
+  expect(successHandler).toHaveBeenCalledTimes(1)
+  expect(successHandler).toHaveBeenCalledWith(undefined)
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+  expect(document.body).toMatchSnapshot()
+
+  await promise
+})
+
+test('it waits for the attributes mutation if configured', async () => {
+  const {container} = render(``)
+
+  const callback = jest
+    .fn(() => container.getAttribute('data-test-attribute'))
+    .mockName('callback')
+  const successHandler = jest.fn().mockName('successHandler')
+  const errorHandler = jest.fn().mockName('errorHandler')
+
+  const promise = waitForElement(callback, {
+    container,
+    mutationObserverOptions: {attributes: true},
+  }).then(successHandler, errorHandler)
+
+  expect(callback).toHaveBeenCalledTimes(1)
+  expect(successHandler).toHaveBeenCalledTimes(0)
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+
+  await skipSomeTimeForMutationObserver()
+
+  expect(callback).toHaveBeenCalledTimes(1)
+  expect(successHandler).toHaveBeenCalledTimes(0)
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+
+  container.setAttribute('data-test-attribute', 'PASSED')
+  await skipSomeTimeForMutationObserver()
+
+  expect(callback).toHaveBeenCalledTimes(2)
+  expect(successHandler).toHaveBeenCalledTimes(1)
+  expect(successHandler).toHaveBeenCalledWith('PASSED')
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+
+  await promise
+})
+
+test('it throws if timeout is exceeded', async () => {
+  const {container} = render(``)
+
+  const callback = jest.fn(() => null).mockName('callback')
+  const successHandler = jest.fn().mockName('successHandler')
+  const errorHandler = jest.fn().mockName('errorHandler')
+
+  const promise = waitForElement(callback, {
+    container,
+    timeout: 300,
+    mutationObserverOptions: {attributes: true},
+  }).then(successHandler, errorHandler)
+
+  expect(callback).toHaveBeenCalledTimes(1)
+  expect(successHandler).toHaveBeenCalledTimes(0)
+  expect(errorHandler).toHaveBeenCalledTimes(0)
+
+  container.setAttribute('data-test-attribute', 'something changed once')
+  await skipSomeTimeForMutationObserver(200)
```

Comment:
Does this time need to be so long? I'd prefer it to be as short as possible so these tests run quickly :)

Reply:
No it doesn't, will reduce to 50.

Reply:
👉 https://github.com/kentcdodds/dom-testing-library/pull/10/commits/edfb24144f1bc1a34dfa4d3e6e91ff1b45a5d81d